### PR TITLE
RB: convert the ruby ApiGraphs to use IPA labels

### PR DESCRIPTION
**Pull request by erik-krogh**
[Evaluation looks clean](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-7930-9739929__Differences__code-scanning/reports).

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll [hunk 1]
```diff
@@ -306,7 +305,7 @@ module API {
    * shortest length.
    */
   cached
-  private module Impl {
+  module Impl {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ApiGraphs.qll:308]: Why this change?

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll [hunk 2]
```diff
@@ -602,48 +601,135 @@ module API {
...
+          any(DataFlowDispatch::ParameterPosition arg).isKeyword(name)
+        } or
+        MkLabelParameter(int n) {
+          any(DataFlowDispatch::ArgumentPosition c).isPositional(n)
+          or
+          any(DataFlowDispatch::ParameterPosition c).isPositional(n)
+        } or
+        MkLabelBlockParameter()
+
+      /** A label for the member named `prop`. */
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ApiGraphs.qll:634]: `A label for a member, for example a constant.`

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll [hunk 3]
```diff
@@ -602,48 +601,135 @@ module API {
...
+      class LabelMember extends ApiLabel {
+        string member;
+
+        LabelMember() { this = MkLabelMember(member) }
+
+        /** Gets the property associated with this label. */
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ApiGraphs.qll:640]: property -> member name

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll [hunk 4]
```diff
@@ -602,48 +601,135 @@ module API {
...
+        MkLabelParameter(int n) {
+          any(DataFlowDispatch::ArgumentPosition c).isPositional(n)
+          or
+          any(DataFlowDispatch::ParameterPosition c).isPositional(n)
+        } or
+        MkLabelBlockParameter()
+
+      /** A label for the member named `prop`. */
+      class LabelMember extends ApiLabel {
+        string member;
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ApiGraphs.qll:636]: I prefer private fields.

**Review by hvitved — CHANGES_REQUESTED** [review]
Thanks a lot for making this change Erik.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll (hunk 1 above)
Comment (erik-krogh) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:308]: I needed that change in the other languages so I could access private predicates.  
I just copy-pasted the change, but it's not needed here, so I'll revert it.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll [hunk 5]
```diff
@@ -602,48 +601,135 @@ module API {
...
+    /** A label in the API-graph */
+    class ApiLabel extends TLabel {
+      /** Gets a string representation of this label. */
+      string toString() { result = "???" }
+    }
+
+    private import LabelImpl
+
+    private module LabelImpl {
+      newtype TLabel =
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:616]: You are suggesting I move `TLabel` into the `Impl` module?  

If we want them to be cached in the same stage we can also just make the module `cached` and add dependencies between the modules. (Basically the cached stages pattern).

**Review by hvitved — APPROVED** [review]
